sharedbuffers/mapped_struct.py

```diff
@@ -1760,6 +1823,252 @@ def _c_search_hkey_ui32(hkey, pindex, stride0, length, hint):
+        mkey = cython.longlong, ikey = cython.int, imkey = cython.int,
+        lo = cython.size_t, hi = cython.size_t, length = cython.size_t,
+        mid = cython.size_t, mid2 = cython.size_t, stride0 = cython.size_t, hint = cython.size_t,
+        pindex = cython.p_char, skip = cython.size_t)
+    @cython.returns(cython.size_t)
+    def _c_search_hkey_i32(hkey, pindex, stride0, length, hint):
+        hi = length
+        lo = 0
+        ikey = hkey
+        if lo < hi:
+            mid = min(hint, hi-1)
+            imkey = cython.cast(cython.p_int, pindex + stride0 * mid)[0]
+            if imkey < ikey:
+                lo = mid = mid+1
+                skip = 32
+                while skip > 0 and mid + skip < hi:
+                    if cython.cast(cython.p_int, pindex + stride0 * (mid+skip))[0] < ikey:
+                        lo = mid+1
+                        mid += skip
+                        skip *= 2
+                    else:
+                        hi = mid + skip
+                        break
+            elif imkey > ikey:
+                hi = mid
+                skip = 32
+                while skip > 0 and mid > lo + skip:
+                    if cython.cast(cython.p_int, pindex + stride0 * (mid-skip))[0] > ikey:
+                        hi = mid
+                        mid -= skip
+                        skip *= 2
+                    else:
+                        lo = mid - skip
+                        break
+            else:
+                while mid > lo and cython.cast(cython.p_int, pindex + stride0 * (mid-1))[0] == ikey:
+                    mid -= 1
+                return mid
+        while lo < hi:
+            mid = (lo+hi)//2
+            imkey = cython.cast(cython.p_int, pindex + stride0 * mid)[0]
+            if imkey < ikey:
+                lo = mid+1
+            elif imkey > ikey:
+                hi = mid
+            else:
+                while mid > lo and cython.cast(cython.p_int, pindex + stride0 * (mid-1))[0] == ikey:
+                    mid -= 1
+                return mid
+        return lo
+
+    @cython.cfunc
+    @cython.locals(
+        hkey = cython.double,
+        mkey = cython.double,
+        lo = cython.size_t, hi = cython.size_t, length = cython.size_t,
+        mid = cython.size_t, mid2 = cython.size_t, stride0 = cython.size_t, hint = cython.size_t,
+        pindex = cython.p_char, skip = cython.size_t)
+    @cython.returns(cython.size_t)
+    def _c_search_hkey_f64(hkey, pindex, stride0, length, hint):
+        hi = length
+        lo = 0
+        if lo < hi:
+            # First iteration a quick guess assuming uniform distribution of keys
+            mid = min(hint, hi-1)
+            mkey = cython.cast(cython.p_double, pindex + stride0 * mid)[0]
+            if mkey < hkey:
+                # Got a lo guess, now skip-search forward for a hi
+                lo = mid = mid+1
+                skip = 32
+                while skip > 0 and mid + skip < hi:
```

Comment:
Yep. Overflow can and does happen. Making a test for it requires a 32-bit system though.